framework/validators/EmailValidator.php

```diff
@@ -176,4 +176,28 @@ public function getClientOptions($model, $attribute)
 
         return $options;
     }
+
+    /**
+     * @param string $value
+     * @return string|bool returns string if it is valid and/or can be converted, bool false if it can't be converted and/or is invalid
+     * @see https://github.com/yiisoft/yii2/issues/18585
+     */
+    private function validateIdnValue ($value)
+    {
+        $idnTest = $this->idnToAscii($value);
+
+        if ($idnTest === false) {
+            $newPatternExploded = explode('@', $this->pattern, 2);
+            $newPattern = $newPatternExploded[0] . "$/";
+
+            $newFullPatternExploded = explode('@', $this->fullPattern);
+            $newFullPattern = $newFullPatternExploded[0] . '@' . $newFullPatternExploded[1]  ."$/";
+
+            if (preg_match($newPattern, $value) || ($this->allowName && preg_match($newFullPattern, $value))) {
+                return $value;
+            }
+        }
+
+        return $idnTest;
+    }
 }
```

Comment:
```suggestion
    private function idnToAsciiWithFallback($value)
    {
        $ascii = $this->idnToAscii($value);

        if ($ascii === false) {
            $parts = explode('@', $this->pattern, 2);
            $pattern = $parts[0] . "$/";

            $fullPatternParts = explode('@', $this->fullPattern);
            $fullPattern = $fullPatternParts[0] . '@' . $fullPatternParts[1]  ."$/";

            if (preg_match($pattern, $value) || ($this->allowName && preg_match($fullPattern, $value))) {
                return $value;
            }
        }
        
        return $ascii;
    }
```

Reply:
Refactored based on your suggestions

Reply:
Awesome! I think it got way more readable. One last issue (left a comment) and it's good to go 👍